aten/src/ATen/core/op_registration/hacky_wrapper_for_legacy_signatures.h

```diff
@@ -207,12 +208,111 @@ constexpr auto with_explicit_optional_tensors(KernelFunc kernel_func) {
+    // that aren't out arguments.
+    // In the aten::example op mentioned above, this would be
+    //   > schema_nonout_arg_indices = [0, 1, 3]
+    using schema_nonout_arg_indices = guts::iseq::remove_all_by_index_t<std::make_index_sequence<num_parameters>, guts::iseq::reverse_t<OutTensorArgIndices>>;
+
+    // kernel_to_schema_permutation then contains a mapping from argument index in KernelFunc to the corresponding
+    // argument index in the schema.
+    // For the aten::example op, that'll be
+    //  > kernel_to_schema_permutation_indices = [2, 4, 0, 1, 3]
+    // Interpreted as a mapping, this means
+    //  - argument 0 in KernelFunc maps to argument 2 in the schema,
+    //  - argument 1 in KernelFunc maps to argument 4 in the schema,
+    //  - argument 2 in KernelFunc maps to argument 0 in the schema,
+    //  - ...
+    // We can use this as a permutation function to reorder types or values correspondingly
+    using kernel_to_schema_permutation_indices = guts::iseq::concat_t<OutTensorArgIndices, schema_nonout_arg_indices>;
+    using kernel_to_schema_permutation = guts::iseq::permutation<kernel_to_schema_permutation_indices>;
+
+    // We also need the inverse permutation because parameters (i.e. types) and arguments (i.e. values)
+    // need to be mapped in inverted directions. For types, we generate the schema order types from
+    // the KernelFunction types, but for arguments we get schema order arguments and need to generate
+    // the KernelFunction arguments.
+    using schema_to_kernel_permutation = typename kernel_to_schema_permutation::inverted_t;
+
+    using target_signature_parameters = typename schema_to_kernel_permutation::template apply_to_typelist_t<
+        typename kernel_signature_traits::parameter_types>;
+
+    template<class Return, class ParameterList>
+    struct wrapper_;
+    template<class Return, class... Parameters>
+    struct wrapper_<Return, guts::typelist::typelist<Parameters...>> {
+        static Return call(Parameters... args) {
+            // call through to KernelFunc but reorder arguments as determined
+            // by the permutation we calculated above.
+            return guts::apply(KernelFunc::func_ptr(),
+                kernel_to_schema_permutation::apply_to_tuple(std::tuple<Parameters...>(std::forward<Parameters>(args)...)));
+        }
+    };
+
+public:
+    using wrapper = wrapper_<typename kernel_signature_traits::return_type, target_signature_parameters>;
+};
+
+
+/**
+ * Take a kernel function that has a number of `Tensor`, `const Tensor&` or `Tensor&` arguments
+ * where all `Tensor&` arguments are at the beginning, and take a `std::index_sequence OutTensorArgIndices`.
+ * Create a wrapper function that has `Tensor&` arguments at the indices given by `OutTensorArgIndices`
+ * and calls through the underlying kernel function by reordering them to the front.
+ * This wrapper also changes the argument type from `Tensor&` to `const Tensor&`.
```

Comment:
Should probably say why we're doing this

Reply:
added a paragraph.

Reply:
turns out I was wrong, neither is the wrapper doing this step of changing `Tensor&` to `const Tensor&` nor is it necessary. The dispatcher has a workaround hack implemented to support `Tensor&`, see https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/core/boxing/impl/make_boxed_from_unboxed_functor.h#L331